### PR TITLE
Pybind11/windows bugfixs

### PR DESCRIPTION
**Pull request by roigcarlo**
I've been testing the branch in windows and it had some problems. Not all of them related with pybind itself, but anyway.

Most important things I found:
* In windows pybind will not cast double <-> int <-> unsigned int types in the binding interface, so the types must be the same as the ones in the functions, Otherwise its a compile error.
* The fact of splitting the core and the interface has exposed some symbols that previously where hidden and now need to be exported. Luckily the mechanism is the same as before, so just add the KRATOS_API() macro to all the classes you need to use in the python interface if this happens.
* The application needs to be PUBLIC in the CMake, the module can be PRIVATE. If the application core is set to public some symbols will be kept hidden even when explicitly exported.

I fixed the issues for the "kernel" applications, but I may have missed something. Please take a look

### CODE REVIEW

File: applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h
Comment (loumalouomega) [inline applications/ContactStructuralMechanicsApplication/custom_linear_solvers/mixedulm_linear_solver.h:704]: Sorry :+1:

**Review by loumalouomega — APPROVED** [review]

**Comment by loumalouomega** [comment]
Ups sorry I wasnt't assigned as reviewer

**Comment by pooyan-dadvand** [comment]
I'm not sure about the fixes made here. I think we should discuss it a little bit before merging this PR.

**Comment by roigcarlo** [comment]
@pooyan-dadvand Agree. @loumalouomega You can cherrypick the non pybind-related if you want, they have nothing to do with pybind as far as I know... ( b3095d7, 2f879b4, 01be759, a6e5fb2 )

**Comment by loumalouomega** [comment]
I can wait, no worries

**Comment by roigcarlo** [comment]
@pooyan-dadvand Fixed the issue.

**Comment by jcotela** [comment]
@roigcarlo just so that I understand: we want PUBLIC when linking against KratosCore (or a base application) and PRIVATE when defining the python module?

**Comment by roigcarlo** [comment]
Exactly

**Comment by roigcarlo** [comment]
@RiccardoRossi approve after the policy change pls?

**Pull request merged**
